### PR TITLE
Update client IP address safelist doc and app for ASP.NET Core 3.1

### PR DESCRIPTION
**Pull request by scottaddie**
Fixes https://github.com/dotnet/AspNetCore.Docs/issues/17255

**Summary of changes**

* Move middleware and filters to class library to avoid duplication across ASP.NET Core versions
* Replace NLog with console logging
* Incorporate Acrolinx feedback
* Introduce versioned moniker ranges
* Emphasize that action filters apply only to MVC
* Expand explanation of code samples
* Add an **Additional resources** section
* Replace HTTP 401 status codes in middleware and filters with HTTP 403

[Internal review page](https://review.docs.microsoft.com/aspnet/core/security/ip-safelist?view=aspnetcore-3.1&branch=pr-en-us-17281)

### CODE REVIEW

File: aspnetcore/security/ip-safelist.md [hunk 1]
```diff
@@ -1,70 +1,148 @@
...
+The `Startup.Configure` method adds the custom `AdminSafeListMiddleware` middleware type to the app's request pipeline. The safelist is retrieved with the .NET Core configuration provider and is passed as a constructor parameter.
 
-[!code-csharp[](ip-safelist/samples/2.x/ClientIpAspNetCore/Startup.cs?name=snippet_Configure&highlight=10)]
+::: moniker range=">= aspnetcore-3.0"
 
-The middleware parses the string into an array and looks for the remote IP address in the array. If the remote IP address is not found, the middleware returns HTTP 401 Forbidden. This validation process is bypassed for HTTP Get requests.
+[!code-csharp[](ip-safelist/samples/3.x/ClientIpAspNetCore/Startup.cs?name=snippet_ConfigureAddMiddleware)]
 
-[!code-csharp[](ip-safelist/samples/2.x/ClientIpAspNetCore/AdminSafeListMiddleware.cs?name=snippet_ClassOnly)]
+::: moniker-end
```
Comment (Rick-Anderson) [inline aspnetcore/security/ip-safelist.md]: 10 moniker pairs are too many. Any reason you didn't do the whole moniker versioning we usually use?

Updating to 5 and 5.1 is going to be more difficult.

Comment (scottaddie) [reply aspnetcore/security/ip-safelist.md]: There are 2 LTS versions of the product: 2.1 and 3.1. We need to have docs for both. I could make this doc available only for 2.1+. Thoughts on that approach?

Comment (Rick-Anderson) [reply aspnetcore/security/ip-safelist.md]: This is what we've been doing:

```
::: moniker range=">= aspnetcore-3.0"
// 3.1
::: moniker-end

::: moniker range="<= aspnetcore-3.0"
// 2.1
::: moniker-end
```

It's pretty easy to put it in that format and easy to test.

Comment (Rick-Anderson) [reply aspnetcore/security/ip-safelist.md]: At this point, we'd consolidate the monikers in another PR. This is how we've been working to update docs to 3.1. Update the doc with 2 sections, but don't update the content:

```
::: moniker range=">= aspnetcore-3.0"
//  copy of 2.1, possibly with consolidation
::: moniker-end

::: moniker range="<= aspnetcore-3.0"
// original 2.1
::: moniker-end
```

Then merge the PR into master/live.  The next PR you update the 3.1 section so the diff is easier to follow for the dev team.

Comment (scottaddie) [reply aspnetcore/security/ip-safelist.md]: I've reduced the number of moniker pairs to 4.

We should discuss the versioning strategy offline. Duplicating content in the manner you've proposed above seems like a bad idea.

**Comment by HaoK** [comment]
I'm probably not the best person to review this, @blowdart or maybe someone from mvc? @mkArtakMSFT

**Comment by blowdart** [comment]
Ugh, no I, for I'd say "Never trust IP addresses". Someone from MVC please.

**Comment by scottaddie** [comment]
@anurse Can someone on your team please review?

**Comment by anurse** [comment]
What's the timeline needed on this? We're pretty slammed right now. Also most of my team is less familiar with MVC, it would be @mkArtakMSFT 's team that has more MVC experience (but they're pretty slammed too).

**Comment by scottaddie** [comment]
@anurse It's not super urgent. We'll wait for someone on @mkArtakMSFT's team to review.

**Pull request merged**